slackapi/bolt-js
# feat: expose auto acknowledgement flag on the `function` handler

### Summary

This many changes were not expected when I first started working on this. If the PR is to large or to hard to review let me know. This PR could be broken into a PR for the refactor of the `function` handler and another for the addition of an auto acknowledgement flag on the `function` handler

#### Changes

This PR changes the way to `event` auto acknowledgement is implemented in this project. These changes lift the auto ack logic into a middleware that is then registered to the proper event listener

This allows the `app.function` listener to expose a `autoAcknowledge` flag that can be used to disable the auto ack middleware. This is required for the Dynamic input work, since functions used as dynamic inputs must `complete` or `fail` before acknowledging the request (this provides a crisp user experience).

#### Example
```ts
app.function('get-options', { autoAcknowledge: false }, async ({ inputs, ack, fail, complete }) => {
  try {
    const category: string[] = inputs.category as string[];
    if (category.includes('dogs')) {
      complete({
        outputs: {
          options: [
            { key: 'Poodle', value: 'value-0' },
            { key: 'Bulldog', value: 'value-1' },
            { key: 'Beagle', value: 'value-2' },
          ],
        },
      });
    } else if (category.includes('cars')) {
      complete({
        outputs: {
          options: [
            { key: 'Toyota', value: 'value-0' },
            { key: 'Ford', value: 'value-1' },
            { key: 'Jeep', value: 'value-2' },
          ],
        },
      });
    } else {
      fail({ error: `Invalid category: ${category} (valid: 'dogs', 'cars')` });
    }
  } catch (error) {
    fail({ error: `Failed to handle a function request (error: ${error})` });
  } finally {
    await ack();
  }
});
```

#### Manual tests

These changes were manually tested against
- [bolt-js-custom-step-template](https://github.com/slack-samples/bolt-js-custom-step-template) -> no breaking change observed 🟢 
- [bolt-ts-starter-template](https://github.com/slack-samples/bolt-ts-starter-template) -> no breaking change observed 🟢 

#### Testing
1. Pull this branch
2. `npm install` & `npm run build`
3. Use `npm install path/to/the/bolt-js/project` to install the built package in a project with a custom step like [bolt-js-custom-step-template](https://github.com/slack-samples/bolt-js-custom-step-template)
4. Set the options field of the function handler to  `{ autoAcknowledge: false }`
5. You should now be able to use the `ack` input parameter in your handler
6. AutoAck behavior is now disabled you must now manually call `await ack();` for this specfic handler

#### Feedback

I've messed around with the `CustomFunction exported types` I am not if these are considered "breaking changes" in this project. There may be a way to work around this.

Should there be more unit tests around these changes?

### Requirements (place an `x` in each `[ ]`)
